### PR TITLE
Require 'basic_debug' priv to view gameplay-relevant debug info (2nd try), require 'debug' priv to view wireframe

### LINKED ISSUE #7245: Control F5 debug info shown to user via server
##### Issue type

 - Feature request

We already have a priv for wireframe but this should perhaps be a HUD flag instead.
>
>Much information is harmless to know: MT version, seed, FPS, view_range, range_all, profiler graphs etc. I see no harm in allowing players to access those, they are very useful for players and are not dev-only information, and being able to access these does not give a bad impression at all.
>
>Perhaps a new HUD flag should control accessing position / direction / pointed nodename.

Source: @paramat at https://github.com/minetest/minetest/issues/4556#issuecomment-324456399

I would like that feature too, that will be nice to have for making survival harder on some servers (with position info allowed by default). Some relevant discussion here too https://github.com/minetest/minetest/issues/4556

### PR DESCRIPTION
**Pull request by Wuzzy2**
Fixes #7245.

This is a revival of PR #8951, but with these changes:

* Profiler screen is no longer touched (it was a huge mistake to add this completely unrelated feature in the PR)
* Added `basic_debug` priv

## What this PR does

* Hide the position, yaw, pitch, mapseed and pointed thing if player does not have the `basic_debug` priv. In practice, the 2nd line of the debug text at the top will disappear and also the entity info when you point an entity. Additionally, the block bounds can't be used as well
* Add `basic_debug` priv to unlock various debug info that could be considered a gameplay advantage if they were accessible by default. NOT granted by default
* With the `debug` priv (which is not new) you also unlock the wireframe

## Justification

The F5 debug screen contains various "spoilers", e.g. info that can give you a gameplay advantage for free: position, pointed node, pointed entity (including health and armor groups). This completely ruins mods like `orienteering` and `compassgps` in which you can craft items that expose your position and more. The mods get ruined because there's no point in crafting a compass if you can get the yaw/position/whatever “for free” with a single key press, and it's unfair to force players to resist the temptation. Players complained, as evidenced by #7245, which also has core dev support.

This PR gets rid of those spoilers unless you have `basic_debug` priv. `orienteering`-style mods can now be created with peace-of-mind.

## To do

Nothing, for now.

## How to test

* Join any world
* Press F5 while you have one of the 4 possible combinations of “you have / have not `debug` priv” and “you have / have not `basic_debug` priv”.

## Expected questions and counter-arguments

1. “This PR is pointless because it's the player's fault for pressing F5.” This argument is bad because it's not good to force players to resist the temptation to enable a cheat that is only 1 keypress away. You could also press F5 by accident, and you can't “unsee” the coordinates. This is not good if a key element of a game is to “earn” the privilege to learn your coordinates through gameplay
2. “Why not using HUD flags?”. HUD flags are controlled by mods. This implies that mods can screw around with the debug info at will. All it takes is one poorly-coded mod to strip away the extended debug info from the admin, which is not fun to track down. Privileges, by comparison, are much more reliable and predictable and there are no ugly surprises.
3. “But how do I get the old behavior back?”: Type `/grantme basic_debug`.
4. “My game/mod/server depends on the pos/angle/pointed thing be visible in debug screen, this PR ruins everything!”. It was never a good idea for a mod to depend on anything written in the debug screen. As the name suggests, its for debugging primarily. The fact that it's currently being abused heavily for gameplay as well is just an excuse for bad practice. Instead of depending on the debug screen, add a mod that exposes the desired info itself, there are countless mods that do this which also can do this with a much nicer HUD than the debug screen. “Knowing the coordinates” is a gameplay advantage and should be considered as such. But you don't need to install any mods, `/grantme basic_debug` still works.

### CODE REVIEW

**Comment by style-nine** [comment]
I have to admit that I don't really see why this doesn't use HUD flags. If an admin is surprised by what a mod is doing, then doesn't that mean that either the mods are too complicated to police properly or that the admins aren't as alert as they should be?

**Comment by Wuzzy2** [comment]
Mods should have NO business in touching the debug info at all. Debug info must work without mods, it's a very essential feature. Allowing mods to screw around with this just makes things unneccessarily complicated. I chose privileges precisely because they are straight-forward, work out of the box and are reliable, no matter what crazy tower of mods you have installed. Also, you failed to provide any argument *for* HUD flags.

**Comment by style-nine** [comment]
It may be debug info, but it uses the HUD to present it. Using HUD flags is consistent with the game as it's already implemented. Treating the mods like an unknown and a danger is in itself a dangerous habit to reinforce that is totally unecessary. Mods are entirely within the control of the server admins. Mods themselves can regulate other mods. You can always implement a master mod that governs the others, and privs can always be circumvented with unknown mods, so I don't see how privs is a fix where mods are not.

**Comment by LoneWolfHT** [comment]
Unless support was added for mods using debug info to display their own values I see no reason for mods to control this.
If you want all your players to see their coordinates on their HUD it's not hard to make a mod to do that, which is probably what we should have been doing in the first place

File: builtin/game/privileges.lua
Comment (paramat) [inline builtin/game/privileges.lua]: true

**Comment by paramat** [comment]
I like how this moves the wireframe viewing priv to a more suitably named priv 'wireframe'. This then frees up the 'debug' priv for another function.

As requested by 2 core devs in the first PR, the default for players of this new 'debug' priv for viewing position, yaw, pitch, mapseed and pointed thing should be true, currently it is false.
Other than that i have no objections.

File: builtin/game/privileges.lua
Comment (paramat) [inline builtin/game/privileges.lua]: 'Can enable' is awkward English and has a double meaning, better is "Allows viewing wireframe mode".

Comment (paramat) [reply builtin/game/privileges.lua]: Oh ... just seen many priv descriptions use "Can ...", that is bad =)

**Comment by Wuzzy2** [comment]
Sadly, granting `debug` to singleplayer by default means you can't get rid of it in singleplayer, as all privs granted by default are irrevokable. This would defeat the very purpose of this PR. This is unacceptable!

**Comment by Wuzzy2** [comment]
Also, see the FAQ above: Just type in `/grantme debug` and you'll get back the old behavior. Or maybe even add it into your default privs in `minetest.conf`.

**Comment by paramat** [comment]
> Sadly, granting debug to singleplayer by default means you can't get rid of it in singleplayer, as all privs granted by default are irrevokable. This would defeat the very purpose of this PR.

Oh hmm ... i did not know this, sorry, removing my -1.
However, the objection from core devs remains, so this seems to be a dilemma.

**Comment by style-nine** [comment]
So, the inability to revoke a priv in singleplayer if it's default would be a bug, don't you think?

**Comment by paramat** [comment]
I agree, we should consider changing that.

**Comment by rubenwardy** [comment]
> So, the inability to revoke a priv in singleplayer if it's default would be a bug, don't you think?

You can revoke default privileges in singleplayer.

There's a feature called `give_to_singleplayer` which means that singleplayer always has certain privileges such as interact, hence why they can't be removed

**Comment by rubenwardy** [comment]
The changing of the meaning of `debug` will cause older clients to be able to use the wireframe if they are given that privilege

**Comment by paramat** [comment]
Thanks.
Yes, looks like 'debug' has to remain for wireframe, and a new priv name added.

**Comment by Wuzzy2** [comment]
I don't understand. How can the player possibly revoke their default privs?

**Comment by rubenwardy** [comment]
minetest.conf:

```
default_privs = fly,fast,noclip
```

in game: `/revoke singleplayer all`

**Comment by Wuzzy2** [comment]
Sigh. OK, maybe the irrevokable privs can be fixed in a different PR then.

Just to be clear: With this change, `debug` will become an irrevokable priv in singleplayer.

File: builtin/game/privileges.lua
Comment (paramat) [inline builtin/game/privileges.lua]: Due to
```
		if def.give_to_singleplayer == nil then
			def.give_to_singleplayer = true
		end
		if def.give_to_admin == nil then
			def.give_to_admin = def.give_to_singleplayer
		end
```
Both lines can be removed.

File: builtin/game/privileges.lua
Comment (paramat) [reply builtin/game/privileges.lua]: So, i am not too bothered about this being improved, up to you. We could improve all at once later.

**Comment by paramat** [comment]
> The changing of the meaning of debug will cause older clients to be able to use the wireframe if they are given that privilege

Unfortunately 'debug' has to remain for wireframe, and a new priv name added for position/direction/seed/pointed.

**Comment by Wuzzy2** [comment]
@paramat: I have removed the two lines.
Sadly, I do not know how to proceed.

File: builtin/game/privileges.lua
Comment (paramat) [reply builtin/game/privileges.lua]: Done.

**Comment by paramat** [comment]
Seems to me the following is necessary:

The 'debug' priv must continue to control wireframe, for the reason explained above.
The description of 'debug' must be changed to be clear it is for wireframe mode only.
We need to choose a new name for your new priv that controls 'basic debug text'.

File: builtin/game/privileges.lua
Comment (paramat) [reply builtin/game/privileges.lua]: Comment now invalid as privs need changing.

**Comment by Wuzzy2** [comment]
Using the `debug` name exclusively for wireframe is a little confusing name, however. Just teach server operators to update their privs.

I am not really happy with this proposal. :-(

Is there no better solution for this?

**Comment by paramat** [comment]
Yeah it is messy and unfortunate, but as 'debug' is granted by default and usually granted, all older clients will then be able to use wireframe mode, causing cheating on servers.
Even if we try to inform, many servers will simply not make the necessary changes, the risk seems too big.

**Comment by Wuzzy2** [comment]
Oh crap. Auto-granting wireframe to everyone is not good. Yeah, you got a point.

**Comment by Wuzzy2** [comment]
New changes!

* `debug` priv is no for wireframe only (old behavior) (renamed from `wireframe`)
* New priv: `basic_debug` for enabling displaying gameplay-relevant debug info text (renamed from `debug`)
* The `wireframe` priv is gone
* Both privs are `give_to_singleplayer = false`

**Comment by paramat** [comment]
The 'debug' priv (wireframe) should also have `give_to_admin = true,`, as before.
The 'basic_debug' priv should be granted by default, as discussed earlier. Priv name seems fine to me.

**Comment by Wuzzy2** [comment]
Hopefully I did it right this time.

File: builtin/game/privileges.lua
Comment (paramat) [inline builtin/game/privileges.lua]: Line unnecessary due to default behaviour.

**Comment by paramat** [comment]
Fine, apart from an unnecessary line.

**Comment by Wuzzy2** [comment]
Done!

**Comment by BluebirdGreycoat** [comment]
Status of this PR? My server has a custom coordinate system (offset from the real coords, with multiple realms), and being able to see the debug coordinates often causes confusion for new players, plus it tends to break the immersion. Would be nice to be able to disable it and ensure everyone uses the same custom coordinate system.

**Comment by Wuzzy2** [comment]
Status: Waiting for (re-)review.

**Comment by Warr1024** [comment]
This is one of the top engine issues that's been an obstacle to NodeCore, with no workaround possible, and it has a notable impact on both singleplayer and multiplayer, so I'm looking forward very much to seeing it merged.

**Comment by Wuzzy2** [comment]
Bump after 5 months.

**Comment by appgurueu** [comment]
`basic_debug` and `debug` privs should be granted by default. This can't be done using current settings, as those are already set on older servers and there should be no breaking changes.

:+1: for the feature, :-1: for breaking compat

**Comment by Wuzzy2** [comment]
How does this break compat?
I disagree that they should be granted by default.
To admin: Maybe, but probably not. To singleplayer: Definitely not. Why? Because most players are … well … players, and not devs. Therefore, disabling it is the more reasonable default.

But even if I would agree, the problem is: The only way to currently grant anything by default (via the engine) is by setting (or not setting) one of the `give_to_singleplayer` or `give_to_admin`. The problem here is that ALL privs granted by default are **irrevokable**. This would be a serious annoyance and defeats the purpose of this PR. Especially for singleplayer

Besides, there is a setting somewhere in which you can already change your initial privileges. I think this covers your use case just fine, so no changes to the code are neccessary after all. :P

**Comment by appgurueu** [comment]
No. That setting does not apply to already joined players and would need to be changed by admins. Many players rely on debug for coordinates.

**Comment by BluebirdGreycoat** [comment]
Bump. I really want to be able to turn debug coordinates off.

**Comment by GreenXenith** [comment]
Agreed, this is a very necessary step towards giving mods/games better control.
Are there any issues remaining to solve?

**Comment by appgurueu** [comment]
I <small>(now)</small> support this. Wireframe etc. is not gameplay relevant so it makes sense to disable it by default. It would be nice to keep backwards compatibility for basic_debug, having it default to true and requiring it to be explicitly disabled, but in order to achieve this, dirty hacks would be needed, such as having a basic_debug_disabled priv, which definitely is a worse solution.

Requiring admins to overthink whether they want their users to be able to have permanent access to coords is therefore justified.

**Comment by hecktest** [comment]
minetest_game can grant players basic_debug on join, everyone else should be forced to make a decision whether they want the spoiler hud back. I sure won't miss it.

Problem is, any punk with a debugger can obtain this info anyway as long as it is sent to the client.

**Comment by appgurueu** [comment]
> Problem is, any punk with a debugger can obtain this info anyway as long as it is sent to the client.

Punks without debuggers can too. This unfortunately is nontrivial to fix.

1. Not sending absolute positions to the client would greatly complicate things.
2. Not sending obstructed chunks & objects within them would require significant server effort & nontrivial visibility tests.

**Comment by lhofhansl** [comment]
In the end I would have been happier with a "debug" and "wireframe" privs. But since that is not apparently possible this is good.

I do *not* think that these privs should be granted to anyone by default. If you're an "admin" and cannot figure our to grant these privs to yourself then ... 

So +1 on this in concept. I'll try it out today.

**Comment by lhofhansl** [comment]
So if I'm the admin of the server should I be able to revoke the privs or not? Right now I get them by default and I cannot revoke them.

**Comment by Wuzzy2** [comment]
Okay, I should clarify: The point of this PR is !!!NOT!!! to implement an effective anti-cheat. I am aware that hiding the coordinates is not enough if we want to seriously hide this info. I

It's more about letting games give control about whether they want to give the player coordinates for free (it's only 1 keypress away, after all) or if they have to earn it (e.g. by crafting a tool or something).
Right now, F5 acts like an "1-click cheat" for all games that do not want to give you the coordinates for free.

**Comment by lhofhansl** [comment]
@Wuzzy2 totally agree. I agree with what you want to do, and it's a shame it takes almost a year to do that.

When I tested it, though, both admins and singleplayer still have basic_debug by default (and that cannot be revoked), and I thought that is what you wanted to avoid.

**Comment by appgurueu** [comment]
I would like to request an addition: a priv `disable_hud` for F1 (disabling the HUD), as that can be used to evade gameplay effects such as flash grenades. Disabling particles should also be controllable by the server.

**Comment by Wuzzy2** [comment]
> I would like to request an addition: a priv disable_hud for F1 (disabling the HUD), as that can be used to evade gameplay effects such as flash grenades. Disabling particles should also be controllable by the server.

This is unrelated to this PR. Besides, I think it would be better to be able to force certain HUD elements to be unhidable, rather than disallowing to disable to hide the whole HUD.

**Comment by Wuzzy2** [comment]
This PR has become old and dusty, so I have updated and rebased it to 5.4.0-dev.

I have also removed giving the privileges to anyone by default, as requested before.

**Comment by daroberts00** [comment]
As my game/server nears completion, personally it frustrates me greatly the extra work I have to put in to quests etc in a survival/semi-RPG server just because the coordinates are there. Wuzzy has a mod that exposes these values if a player crafts the items, but then if the server owner wants the game to be even more "hardcore" - there's always the sun! It should be up to the game designer how their game works, and the server owner what info is provided to the player.

**Comment by celeron55** [comment]
This was and is still a good idea. I believe rubenwardy will hit the merge button soon enough if this still works.

**Comment by rubenwardy** [comment]
If you grant/revoke `basic_debug` when the debug interface is open, nothing changes. It should probably trigger an update when that privilege is modified

**Review by rubenwardy — DISMISSED** [review]

**Comment by rubenwardy** [comment]
Okay, to summarise that way too long IRC discussion: the priv is fine and so is removing it by default too.

The only changes required are:

1. As mentioned above, an update check to fix the bug with priv changes
2. A protocol version check to not require the priv on 5.4 or lower servers

**Comment by Wuzzy2** [comment]
This PR just got rebased ... again.

There are 2 new commits.

I have finally fixed the issue that the debug screen does not update on `basic_debug` priv change. The debug screen will now update instantly as required.

Also, I have applied a new restricion to the "mapblock bounds" feature with F8. This key does not work unless you have the `basic_debug` privilege. The reason is that knowing where the mapblock boundaries are might also be considered a gameplay advantage (even if a minor). Also, it's technically also a debug thing. This is me just following the spirit of #7245.

**Comment by Wuzzy2** [comment]
Protocol version check added but not tested by me. **This PR now bumps the protocol to 40!**

Thanks to Warr1024.

File: src/network/clientpackethandler.cpp
Comment (rubenwardy) [inline src/network/clientpackethandler.cpp:900]: missing space, maybe a comment would be good too

**Review by rubenwardy — APPROVED** [review]

**Comment by rubenwardy** [comment]
Note: I tested everything except the protocol version bump

**Comment by Warr1024** [comment]
I have tested this and it works as expected, and the code looks fine to me, though I don't feel confident enough to give it the final merging approval.

- 5.5 client on 5.5 server: works, selected info is hidden/unavailable, granting and revoking the priv behaves as expected.
- 5.5 client on 5.4 server: works, selected info is still shown as per the existing 5.4 behavior.
- 5.4 client on 5.5 server: not applicable as the changes only affect the 5.5+ client.

**Comment by celeron55** [comment]
I see this was tested to work and the code looks fine to me.

**Comment by appgurueu** [comment]
Yay

**Comment by sharedrory** [comment]
Not cool. Soon everything with require a priv. Respawning, looking around, jumping, clicking, audio, joining a serv-

**Comment by Wuzzy2** [comment]
@sharedrory: Slippery slope fallacy.

**Comment by appgurueu** [comment]
> @sharedrory: Slippery slope fallacy.

Slippery slope? I'd rather call this fallacy "taking it to the extremes".

These debug features should never have been included in non-debug clients in the first place.

**Comment by sharedrory** [comment]
> @sharedrory: Slippery slope fallacy.

I don't think you've considered the number of servers that DON'T use those mods. Now we have to go bug the server owners to give us the priv. The fact that you don't get it in singleplayer means you now have to type out yet another 2 commands to get it unless you want to grant all at the same time.

> These debug features should never have been included in non-debug clients in the first place.

Then atleast have mtg add a hud element with the coords? Perhaps a system like the minimap where you need the mapping kit. Speaking of the mapping kit, it reveals the direction you are facing (north, south, ...).

**Comment by appgurueu** [comment]
> I don't think you've considered the number of servers that DON'T use those mods. Now we have to go bug the server owners to give us the priv.

You should not bug them to give you the priv. You should bug them to install a mod which provides the same essential functionality (coords, I assume).

**Comment by sharedrory** [comment]
> You should **not bug** them to give you the priv. You should **bug** them to install a mod which provides **the same** **_essential_** functionality (coords, I assume).

Why don't we make the engine available behind a priv. And then let a mod provide the same essential functionality.

**Comment by sfan5** [comment]
If it's widely considered not to be a problem then Minetest Game could also grant this priv by default. Other games would still have the freedom not to do this and users will be happy that they don't have to bother server admins for this.

**Comment by appgurueu** [comment]
> If it's widely considered not to be a problem then Minetest Game could also grant this priv by default. Other games would still have the freedom not to do this and users will be happy that they don't have to bother server admins for this.

The only information that really is must be displayed are the coordinates, which are relied on for navigation. The other clutter isn't needed.

**Pull request merged**
